File: apps/desktop/src/types/pomodoro.ts
export type PomodoroPhase = "focus" | "shortBreak" | "longBreak";
export type PomodoroControlState = "idle" | "running" | "paused";
export type PomodoroSessionOutcome = "completed" | "interrupted" | "skippedBreak";
export type PomodoroTransitionKind =
  | "sessionStarted"
  | "paused"
  | "resumed"
  | "focusCompleted"
  | "breakStarted"
  | "breakCompleted"
  | "sessionStopped"
  | "breakSkipped"
  | "nextFocusStarted";

export interface PomodoroPreset {
  label: string;
  focusMinutes: number;
  shortBreakMinutes: number;
  longBreakMinutes: number;
  sessionsUntilLongBreak: number;
}

export interface PomodoroSnapshot {
  controlState: PomodoroControlState;
  phase: PomodoroPhase | null;
  preset: PomodoroPreset;
  sessionStartedAt: string | null;
  phaseStartedAt: string | null;
  phaseEndsAt: string | null;
  pausedAt: string | null;
  remainingSeconds: number;
  phaseTotalSeconds: number;
  phaseElapsedSeconds: number;
  focusSecondsElapsed: number;
  breakSecondsElapsed: number;
  completedFocusBlocks: number;
  completedBreaks: number;
  autoStartBreaks: boolean;
  autoStartFocus: boolean;
  canPause: boolean;
  canResume: boolean;
  canStop: boolean;
  canSkipBreak: boolean;
  sessionId: number | null;
  outcome: PomodoroSessionOutcome | null;
}

export interface PomodoroTransition {
  id: number;
  kind: PomodoroTransitionKind;
  title: string;
  body: string;
  state: PomodoroSnapshot;
}

export interface PomodoroEvent {
  state: PomodoroSnapshot;
}

export interface StartPomodoroRequest {
  label: string;
  focusMinutes: number;
  shortBreakMinutes: number;
  longBreakMinutes: number;
  sessionsUntilLongBreak: number;
  autoStartBreaks: boolean;
  autoStartFocus: boolean;
}

export const defaultPomodoroPreset: PomodoroPreset = {
  label: "Classic",
  focusMinutes: 25,
  shortBreakMinutes: 5,
  longBreakMinutes: 15,
  sessionsUntilLongBreak: 4,
};

export const defaultPomodoroSnapshot: PomodoroSnapshot = {
  controlState: "idle",
  phase: null,
  preset: defaultPomodoroPreset,
  sessionStartedAt: null,
  phaseStartedAt: null,
  phaseEndsAt: null,
  pausedAt: null,
  remainingSeconds: 0,
  phaseTotalSeconds: 0,
  phaseElapsedSeconds: 0,
  focusSecondsElapsed: 0,
  breakSecondsElapsed: 0,
  completedFocusBlocks: 0,
  completedBreaks: 0,
  autoStartBreaks: false,
  autoStartFocus: false,
  canPause: false,
  canResume: false,
  canStop: false,
  canSkipBreak: false,
  sessionId: null,
  outcome: null,
};

File: crates/focus-domain/src/pomodoro.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::PomodoroPreset;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum PomodoroPhase {
    Focus,
    ShortBreak,
    LongBreak,
}

impl PomodoroPhase {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Focus => "focus",
            Self::ShortBreak => "short_break",
            Self::LongBreak => "long_break",
        }
    }

    pub const fn is_break(self) -> bool {
        matches!(self, Self::ShortBreak | Self::LongBreak)
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum PomodoroControlState {
    Idle,
    Running,
    Paused,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum PomodoroSessionOutcome {
    Completed,
    Interrupted,
    SkippedBreak,
}

impl PomodoroSessionOutcome {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Completed => "completed",
            Self::Interrupted => "interrupted",
            Self::SkippedBreak => "skipped_break",
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum PomodoroTransitionKind {
    SessionStarted,
    Paused,
    Resumed,
    FocusCompleted,
    BreakStarted,
    BreakCompleted,
    SessionStopped,
    BreakSkipped,
    NextFocusStarted,
}

impl PomodoroTransitionKind {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::SessionStarted => "session_started",
            Self::Paused => "paused",
            Self::Resumed => "resumed",
            Self::FocusCompleted => "focus_completed",
            Self::BreakStarted => "break_started",
            Self::BreakCompleted => "break_completed",
            Self::SessionStopped => "session_stopped",
            Self::BreakSkipped => "break_skipped",
            Self::NextFocusStarted => "next_focus_started",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PomodoroSnapshot {
    pub control_state: PomodoroControlState,
    pub phase: Option<PomodoroPhase>,
    pub preset: PomodoroPreset,
    pub session_started_at: Option<DateTime<Utc>>,
    pub phase_started_at: Option<DateTime<Utc>>,
    pub phase_ends_at: Option<DateTime<Utc>>,
    pub paused_at: Option<DateTime<Utc>>,
    pub remaining_seconds: i64,
    pub phase_total_seconds: i64,
    pub phase_elapsed_seconds: i64,
    pub focus_seconds_elapsed: i64,
    pub break_seconds_elapsed: i64,
    pub completed_focus_blocks: u32,
    pub completed_breaks: u32,
    pub auto_start_breaks: bool,
    pub auto_start_focus: bool,
    pub can_pause: bool,
    pub can_resume: bool,
    pub can_stop: bool,
    pub can_skip_break: bool,
    pub session_id: Option<i64>,
    pub outcome: Option<PomodoroSessionOutcome>,
}

impl PomodoroSnapshot {
    pub fn idle(preset: PomodoroPreset) -> Self {
        Self {
            control_state: PomodoroControlState::Idle,
            phase: None,
            preset,
            session_started_at: None,
            phase_started_at: None,
            phase_ends_at: None,
            paused_at: None,
            remaining_seconds: 0,
            phase_total_seconds: 0,
            phase_elapsed_seconds: 0,
            focus_seconds_elapsed: 0,
            break_seconds_elapsed: 0,
            completed_focus_blocks: 0,
            completed_breaks: 0,
            auto_start_breaks: false,
            auto_start_focus: false,
            can_pause: false,
            can_resume: false,
            can_stop: false,
            can_skip_break: false,
            session_id: None,
            outcome: None,
        }
    }

    pub const fn is_running(&self) -> bool {
        matches!(self.control_state, PomodoroControlState::Running)
    }

    pub const fn is_paused(&self) -> bool {
        matches!(self.control_state, PomodoroControlState::Paused)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PomodoroTransition {
    pub id: u64,
    pub kind: PomodoroTransitionKind,
    pub title: String,
    pub body: String,
    pub state: PomodoroSnapshot,
}

pub fn recommended_presets() -> Vec<PomodoroPreset> {
    vec![
        PomodoroPreset::default(),
        PomodoroPreset {
            label: "Deep Work".to_string(),
            focus_minutes: 45,
            short_break_minutes: 10,
            long_break_minutes: 20,
            sessions_until_long_break: 3,
        },
        PomodoroPreset {
            label: "Sprint".to_string(),
            focus_minutes: 60,
            short_break_minutes: 10,
            long_break_minutes: 20,
            sessions_until_long_break: 2,
        },
    ]
}
